geomstats/learning/geometric_median.py

```diff
@@ -1,116 +1,130 @@
         Optional, default : None
+    epsilon : float
+        Tolerance for stopping the algorithm (distance between two successive
+        estimates).
+        Optional, default : gs.atol
+
+    Attributes
+    ----------
+    estimate_ : array-like, shape={representation shape}
+        If fit, geometric median.
 
     References
     ----------
-    .. [BJL2008]_ Bhatia, Jain, Lim. "Robust statistics on
-        Riemannian manifolds via the geometric median"
+    .. [FVJ2009] Fletcher PT, Venkatasubramanian S and Joshi S.
+        "The geometric median on Riemannian manifolds with application to
+        robust atlas estimation", NeuroImage, 2009
+        https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2735114/
     """
 
-    def __init__(self, metric, max_iter=100, lr=1.0, init=None, print_every=None):
+    def __init__(
+        self,
+        metric,
+        max_iter=100,
+        lr=1.0,
+        init=None,
+        print_every=None,
+        epsilon=gs.atol,
+    ):
         self.metric = metric
         self.max_iter = max_iter
         self.lr = lr
         self.init = init
         self.print_every = print_every
+        self.epsilon = epsilon
         self.estimate_ = None
 
     def _iterate_once(self, current_median, X, weights, lr):
-        """Compute a single iteration of Weiszfeld Algorithm.
+        """Compute a single iteration of Weiszfeld algorithm.
 
         Parameters
         ----------
         current_median : array-like, shape={representation shape}
-            current median.
-        X : array-like, shape=[..., {representation shape}]
-            data for which geometric has to be found.
-        weights : array-like, shape=[N]
-            weights for weighted sum.
+            Current median.
+        X : array-like, shape=[n_samples, {representation shape}]
+            Training input samples.
+        weights : array-like, shape=[n_samples,]
+            Weights for weighted sum.
         lr : float
-            learning rate for the current iteration.
+            Learning rate for the current iteration.
 
         Returns
         -------
-        updated_median: array-like, shape={representation shape}
-            updated median after single iteration.
+        updated_median : array-like, shape={representation shape}
+            Updated median after single iteration.
         """
-
-        def _scalarmul(scalar, array):
-            if gs.ndim(array) == 2:
-                return scalar[:, None] * array
-
-            return scalar[:, None, None] * array
-
         dists = self.metric.dist(current_median, X)
 
         if gs.allclose(dists, 0.0):
             return current_median
 
         logs = self.metric.log(X, current_median)
-        mul = gs.divide(weights, dists, ignore_div_zero=True)
-        v_k = gs.sum(_scalarmul(mul, logs), axis=0) / gs.sum(mul)
+        w = gs.divide(weights, dists, ignore_div_zero=True)
+        v_k = gs.einsum("n,n...->...", w, logs) / gs.sum(w)
```

Comment:
This part of the code is not sufficiently robust (not your changes, the approach that was already being used).

`tensorflow` tests are failing because `gs.divide` is returning `[0., 0., x]`, whereas the others return `[huge number, huge number, x]`. the paper tells to ignore points that are the median in the estimation of the gradient, since they're singularities (as usual, this brings numerical issues).

The problem of ignoring points (which is implicitly what `tf` is doing) is that the algorithm is not converging (should it @qbarthelemy for this particular case?). In fact, is going cyclic between two medians.

Do you have any idea how to make this more robust @qbarthelemy, @SaitejaUtpala, @shubhamtalbar96?

Reply:
To avoid division by zero, I suggest to not initialize on the last sample, and to not compute weights associated to dists=0.
I will commit a new version.

Reply:
> To avoid division by zero, I suggest:
> 
>     * to not initialize on the last sample but on the Euclidean mean,
> 
>     * to not compute weights associated to dists=0.
> 
> 
> This implementation is used in pyRiemann and it is quite stable. https://github.com/pyRiemann/pyRiemann/blob/master/pyriemann/utils/median.py#L87 I will commit a new version.

@qbarthelemy @lpereira95  I don't think Euclidean mean is appropriate, as it is not closed operation, ex: hypersphere. Now this can cause distance functions to behave weirdly. Instead employ a projection step after taking the mean as a simple fix.

Reply:
>  Instead employ a projection step after taking the mean as a simple fix.

Yes, great idea! However, I know the `metric`, not the `space`, so I can't call function `space.projection(gs.mean(X, axis=0))`.
To do that, we must add a parameter `space` to the class `GeometricMedian`. Is it ok for you? @lpereira95 @SaitejaUtpala

Reply:
ah you're right @qbarthelemy.

I think for now we should keep only `metric` as input (maybe something to be reviewed soon), for consistency with most of the current learning methods.

What we can do is to keep the previous way of initialization and control `init_point` in tests. Maybe this should also be the way a user uses the algorithm.

after this, let's merge.